runtime/cudaq/algorithms/observe.h

```diff
@@ -50,6 +50,18 @@ concept ObserveCallValid =
     ValidArgumentsPassed<QuantumKernel, Args...> &&
     HasVoidReturnType<std::invoke_result_t<QuantumKernel, Args...>>;
 
+/// \brief Observe options to provide to the observe() / async_observe()
+/// functions
+///
+/// \param shots number of shots to run for the given kernel. The default of -1
+/// means direct calculations for simulation backends or platform default for
+/// remote backends.
+/// \param noise noise model to use for the observe operation
+struct observe_options {
+  int shots = cudaq::get_platform().get_shots().value_or(-1);
```

Comment:
I think we should move away from using the shots in the platform, it should be on a per-function call basis. So here I would just set it to -1.

Reply:
If we do that, then this hypothetical code wouldn't work correctly because it would choose direct calculation and not be shots-based, which would seem contrary to the programmer's intent.
```c++
cudaq::set_shots(1000);
cudaq::noise_model noise; /* ...add noise... */
double energy = observe({.noise = noise}, ansatz, h, 0.59);
```
If we don't want to support that, I feel like we should deprecate these functions in lieu of this new preferred interface?
```c++
cudaq::set_shots();
cudaq::clear_shots();
cudaq::quantum_platform::set_shots();
cudaq::quantum_platform::clear_shots();
```
Thoughts?

Reply:
Yea I'm saying that type of code should be deprecated. That was from the early days, now we don't want to rely on shots coming from global state, but rather have the programmer be intentional at the function callsite.

Reply:
How about a `std::optional<unsigned>` type? Instead of encoding a magic value, `-1`?

Reply:
`-1` was used in multiple places for the `observe` capability, but I'll look to see if they can all be updated to comprehend the `std::optional<std::size_t>` type.

Reply:
I think we had an Issue Tracker up for this from a long time ago - we do need to update the `int` shots for observe to a `std::size_t`. `std::optional` is probably a good idea because we do want to differentiate between shots-based and non-shots based sampling for expectation values.

Reply:
You might want to leave that for a follow on PR

Reply:
Yeah, I took a look, and these changes start compounding and I think would be best left for a separate PR. I think we need to stay with `int` for now, too, because `-1` is needed to denote analytic computation rather than shots-based computation.

Reply:
Deprecated `set_shots` and `clear_shots` methods in [c445f00](https://github.com/NVIDIA/cuda-quantum/pull/664/commits/c445f00bf404443ca5e03c8d243cb1e7293769a8). I think that resolves this conversation? Please let me know if you think otherwise.

Also, comments regarding the meaning of defaults were later updated (to remove the part that said it would honor the platform's shots) in [7e0eeeb](https://github.com/NVIDIA/cuda-quantum/pull/664/commits/7e0eeebbe62b4db93d0dbb38dee0e3d0cdee3c8f).